lib/Sema/SemaExpr.cpp

```diff
@@ -12490,36 +12495,69 @@ ExprResult Sema::ActOnBoundsCastExpr(Scope *S, SourceLocation LParenLoc,
     }
   }
 
-  return BuildBoundsCastExpr(LParenLoc, castTInfo, RParenLoc, E1,
-                             dyn_cast<BoundsExpr>(bounds.get()), kind);
+  return BuildBoundsCastExpr(OpLoc, Kind, castTInfo,
+                             SourceRange(LAngleBracketLoc, RAngleBracketLoc),
+                             SourceRange(LParenLoc, RParenLoc), E1,
+                             dyn_cast<BoundsExpr>(bounds.get()));
 }
 
-ExprResult Sema::GenerateBoundsExpr(Expr *E2, Expr *E3) {
+ExprResult Sema::GenerateBoundsExpr(Expr *E1, Expr *E2, Expr *E3,
+                                    QualType DestTy, int SyntaxType,
+                                    tok::TokenKind Kind,
+                                    SourceLocation TypeLoc) {
   ExprResult Result(true);
   ExprResult ResE2 = CorrectDelayedTyposInExpr(E2);
   ExprResult ResE3 = CorrectDelayedTyposInExpr(E3);
 
-  if (!ResE2.isUsable() && !ResE3.isUsable()) {
-    llvm::APInt I = llvm::APInt(1, 1, false);
-    uint64_t Bits = I.getZExtValue();
-    unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
-    llvm::APInt ResultVal(Width, Bits);
-    IntegerLiteral *One = IntegerLiteral::Create(
-        Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             One, SourceLocation());
-  } else if (ResE2.isUsable() && !ResE3.isUsable()) {
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             ResE2.get(), SourceLocation());
-  } else if (ResE2.isUsable() && ResE3.isUsable()) {
-    Result = ActOnRangeBoundsExpr(SourceLocation(), ResE2.get(), ResE3.get(),
-                                  SourceLocation());
+  switch (SyntaxType) {
+  case BoundsCastExpr::SyntaxType::Single:
+    if (DestTy->isCheckedPointerPtrType() || DestTy->isUncheckedPointerType()) {
+      llvm::APInt I = llvm::APInt(1, 1, false);
+      uint64_t Bits = I.getZExtValue();
+      unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
+      llvm::APInt ResultVal(Width, Bits);
+      IntegerLiteral *One = IntegerLiteral::Create(
+          Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
+      Result =
+          ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
+                               One, SourceLocation());
+    } else if (DestTy->isCheckedPointerArrayType() ||
```

Comment:
You should just fall into this else case.  There should be no check.    The problem is that if DestTy isn't a checked pointer array type or an integral type, the code just fall into the last case without an error.  As I explained in my prior review, for any code that creates an invalid bounds expression, there must always be an error message issued so that we don't generate code.